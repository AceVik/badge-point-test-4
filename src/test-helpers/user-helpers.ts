import { User } from '../types/user.interface';
import { UserWithBadge } from '../types/user-with-badge.interface';
import { getUsersBadge } from '../badge-helpers';

export const getUserMock = (count: number): User => {
  return {
    id: '___',
    username: '___',
    solutionCount: count,
  };
};

export const getUserMockWithBadge = async (count: number): Promise<UserWithBadge> => {
  const user = getUserMock(count);
  return {
    ...user,
    badge: await getUsersBadge(user),
  };
};