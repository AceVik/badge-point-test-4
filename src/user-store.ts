import { faker } from '@faker-js/faker';
import { User } from './types/user.interface';
import { UserWithBadge } from './types/user-with-badge.interface';
import { getUsersBadge } from './badge-helpers';

export const getAllUser = async (): Promise<User[]> => {
  const userCount = getRandomInt(5000);
  return Array.from(Array(userCount), () => generateUser());
};

export const getAllUsersWithBadge = async (): Promise<ReadonlyArray<UserWithBadge>> => {
  const usersWithBadge: Array<UserWithBadge> = [];

  for (const user of await getAllUser()) {
    usersWithBadge.push({
      ...user,
      badge: await getUsersBadge(user),
    });
  }

  return usersWithBadge;
};

const generateUser = (): User => {
  return {
    id: faker.string.uuid(),
    username: faker.person.fullName(),
    solutionCount: getRandomInt(2200),
  };
};

export const getRandomInt = (max: number): number => {
  return Math.floor(Math.random() * max);
};
