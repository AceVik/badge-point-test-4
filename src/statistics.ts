import { User } from './types/user.interface';
import { UsersStatistics } from './types/users-statistics.interface';
import { getAllUsersWithBadge } from './user-store';
import { Icon } from './types/icon.enum';

const sortUsersBySolutionCountDesc = (lhs: User, rhs: User) => rhs.solutionCount - lhs.solutionCount;
const fifthUserIndex = 4;

export const calculateUsersStatistics = async (): Promise<UsersStatistics> => {
  const allUsers = await getAllUsersWithBadge();

  let totalSolutionCount = 0;
  const givenBadgesCounter: Record<Icon, number> = {
    [Icon.BADGE_STARTER]: 0,
    [Icon.BADGE_BRONZE]: 0,
    [Icon.BADGE_SILVER]: 0,
    [Icon.BADGE_GOLD]: 0,
    [Icon.BADGE_PLATINUM]: 0,
    [Icon.BADGE_GOD_LIKE]: 0,
    [Icon.BADGE_BAD_ASS]: 0,
  };

  const topFiveUsers = allUsers.slice(0, 5);
  topFiveUsers.sort(sortUsersBySolutionCountDesc);

  for (let i = 0; i < allUsers.length; i++) {
    const user = allUsers[i];

    totalSolutionCount += user.solutionCount;

    if (i > fifthUserIndex && topFiveUsers[fifthUserIndex].solutionCount < user.solutionCount) {
      topFiveUsers.pop();
      topFiveUsers.push(user);
      topFiveUsers.sort(sortUsersBySolutionCountDesc);
    }

    if (user.badge) givenBadgesCounter[user.badge]++;
  }

  return {
    usersCount: allUsers.length,
    averageSolutionCount: allUsers.length > 0 ?  totalSolutionCount / allUsers.length : 0,
    topFiveUsers,
    mostGivenBadge: Object.entries(givenBadgesCounter).reduce((maxBadge, [badge, count]) => {
      return count > givenBadgesCounter[maxBadge] ? (badge as Icon) : maxBadge;
    }, Object.keys(givenBadgesCounter)[0] as Icon),
  };
};