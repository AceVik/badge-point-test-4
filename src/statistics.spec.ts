import { calculateUsersStatistics } from './statistics';
import { getAllUsersWithBadge } from './user-store';
import { getUserMockWithBadge } from './test-helpers/user-helpers';

jest.mock('./user-store');

const getAllUsersWithBadgeMock = getAllUsersWithBadge as jest.Mock;

describe('calculateUsersStatistics', () => {
  beforeEach(() => {
    getAllUsersWithBadgeMock.mockClear();
  });

  it('should calculate statistics for zero users', async function () {
    getAllUsersWithBadgeMock.mockResolvedValue([]);
    const usersStatistics = await calculateUsersStatistics();
    expect(usersStatistics).toMatchSnapshot();
  });

  it('should calculate statistics for one user', async function () {
    getAllUsersWithBadgeMock.mockResolvedValue([
      await getUserMockWithBadge(55),
    ]);
    const usersStatistics = await calculateUsersStatistics();
    expect(usersStatistics).toMatchSnapshot();
  });

  it('should calculate statistics for five users', async function () {
    getAllUsersWithBadgeMock.mockResolvedValue([
      await getUserMockWithBadge(55),
      await getUserMockWithBadge(4),
      await getUserMockWithBadge(5000),
      await getUserMockWithBadge(4500),
      await getUserMockWithBadge(1400),
    ]);
    const usersStatistics = await calculateUsersStatistics();
    expect(usersStatistics).toMatchSnapshot();
  });

  it('should calculate statistics for more than five users', async function () {
    getAllUsersWithBadgeMock.mockResolvedValue([
      await getUserMockWithBadge(55),
      await getUserMockWithBadge(4),
      await getUserMockWithBadge(5000),
      await getUserMockWithBadge(4500),
      await getUserMockWithBadge(1400),
      await getUserMockWithBadge(2000),
      await getUserMockWithBadge(100),
      await getUserMockWithBadge(1000),
      await getUserMockWithBadge(10000),
      await getUserMockWithBadge(20000),
    ]);
    const usersStatistics = await calculateUsersStatistics();
    expect(usersStatistics).toMatchSnapshot();
  });
});