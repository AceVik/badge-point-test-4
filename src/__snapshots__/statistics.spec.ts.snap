// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateUsersStatistics should calculate statistics for five users 1`] = `
{
  "averageSolutionCount": 2191.8,
  "mostGivenBadge": "badge_god_like",
  "topFiveUsers": [
    {
      "badge": "badge_god_like",
      "id": "___",
      "solutionCount": 5000,
      "username": "___",
    },
    {
      "badge": "badge_god_like",
      "id": "___",
      "solutionCount": 4500,
      "username": "___",
    },
    {
      "badge": "badge_platinum",
      "id": "___",
      "solutionCount": 1400,
      "username": "___",
    },
    {
      "badge": "badge_gold",
      "id": "___",
      "solutionCount": 55,
      "username": "___",
    },
    {
      "badge": "badge_starter",
      "id": "___",
      "solutionCount": 4,
      "username": "___",
    },
  ],
  "usersCount": 5,
}
`;

exports[`calculateUsersStatistics should calculate statistics for more than five users 1`] = `
{
  "averageSolutionCount": 4405.9,
  "mostGivenBadge": "badge_platinum",
  "topFiveUsers": [
    {
      "badge": "badge_god_like",
      "id": "___",
      "solutionCount": 20000,
      "username": "___",
    },
    {
      "badge": "badge_god_like",
      "id": "___",
      "solutionCount": 10000,
      "username": "___",
    },
    {
      "badge": "badge_god_like",
      "id": "___",
      "solutionCount": 5000,
      "username": "___",
    },
    {
      "badge": "badge_god_like",
      "id": "___",
      "solutionCount": 4500,
      "username": "___",
    },
    {
      "badge": "badge_platinum",
      "id": "___",
      "solutionCount": 2000,
      "username": "___",
    },
  ],
  "usersCount": 10,
}
`;

exports[`calculateUsersStatistics should calculate statistics for one user 1`] = `
{
  "averageSolutionCount": 55,
  "mostGivenBadge": "badge_gold",
  "topFiveUsers": [
    {
      "badge": "badge_gold",
      "id": "___",
      "solutionCount": 55,
      "username": "___",
    },
  ],
  "usersCount": 1,
}
`;

exports[`calculateUsersStatistics should calculate statistics for zero users 1`] = `
{
  "averageSolutionCount": 0,
  "mostGivenBadge": "badge_starter",
  "topFiveUsers": [],
  "usersCount": 0,
}
`;
